### ISSUE TITLE
Segmentation fault while rendering a SVG using fill pattern based on SVG image

### ISSUE THREAD

**Issue by tlode** [wkhtmltopdf/wkhtmltopdf#2464]
I am trying to render a page which embeds a SVG using a referenced SVG image in a fill pattern, like in this simplified example:

``` html
<!DOCTYPE html>
<html>
<body>
    <svg xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="300" height="200" version="1.1">
        <defs>
            <pattern id="pattern" x="0" y="0" width="10" height="10" patternUnits="userSpaceOnUse">
                <image xlink:href="/test-pattern.svg" x="0" y="0" width="5" height="5"></image>
            </pattern>
        </defs>
        <g>
            <rect fill="url(#pattern)" x="0" y="0" width="300" height="200" stroke="#000" stroke-width="1"></rect>
        </g>
    </svg>
</body>
</html>
```

and the test-pattern.svg:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="5" height="5">
    <rect x="0" y="0" width="5" height="5" fill="#005599"></rect>
</svg>
```

Running `wkhtmltopdf 0.12.3-dev-79ff51e (with patched qt)` 64bit on OS X, I get a segmentation fault:

``` sh
% /usr/local/bin/wkhtmltopdf 'http://localhost/test.html' test.pdf
Loading pages (1/6)
Counting pages (2/6)
Resolving links (4/6)
Loading headers and footers (5/6)
Printing pages (6/6)
Segmentation fault===========================================] Page 1 of 1
```

There are no problems rendering the SVG in Safari or Chrome. 

When embedding the `test-pattern.svg` code directly within the `<pattern>` tag instead of using `<image xlink:href=... >` there is also no problem.

I've noticed when I omit the xmlns declaration `xmlns="http://www.w3.org/2000/svg"` in `test-pattern.svg` or changing it to `foo`, no seg fault occurs, but the rectangle will be rendered empty. Don't know if this is somehow relevant.

**Comment by alloylab** [comment]
is this still an issue on 0.12.5?

**Comment by tlode** [comment]
Yes, the seg fault still occurs when I run the example from above with the current version, installed by the 64-bit installer on a macos 10.13.3 

```
% wkhtmltopdf -V
wkhtmltopdf 0.12.5 (with patched qt)
```

```
% wkhtmltopdf 'http://localhost/test.html' test.pdf
Loading pages (1/6)
Counting pages (2/6)
Resolving links (4/6)
Loading headers and footers (5/6)
Printing pages (6/6)
Segmentation fault===========================================] Page 1 of 1
```

**Comment by alloylab** [comment]
So you used this package? https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.5/wkhtmltox-0.12.5-1.macos-carbon.pkg

Does this work? wkhtmltopdf 'https://www.google.com' ~/Desktop/google.pdf

Can you also provide details on the localhostst and html file?

**Comment by tlode** [comment]
No, I've installed wkhtmltox-0.12.5-1.macos-cocoa.pkg (from https://wkhtmltopdf.org/downloads.html).

> wkhtmltopdf 'https://www.google.com' ~/Desktop/google.pdf

runs fine.

> Can you also provide details on the localhostst and html file?

Native macOS Apache/2.4.28. The html file was exactly the same as from the example above:

test.html
```html
<!DOCTYPE html>
<html>
<body>
    <svg xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg" width="300" height="200" version="1.1">
        <defs>
            <pattern id="pattern" x="0" y="0" width="10" height="10" patternUnits="userSpaceOnUse">
                <image xlink:href="/test-pattern.svg" x="0" y="0" width="5" height="5"></image>
            </pattern>
        </defs>
        <g>
            <rect fill="url(#pattern)" x="0" y="0" width="300" height="200" stroke="#000" stroke-width="1"></rect>
        </g>
    </svg>
</body>
</html>
```

test-pattern.svg
```svg
<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="5" height="5">
    <rect x="0" y="0" width="5" height="5" fill="#005599"></rect>
</svg>
```

**Comment by alloylab** [comment]
Thanks for confirming so I tried to quickly recreate the issue on my Mac... and I received no seg fault with this command: wkhtmltopdf 'http://localhost/test.html' ~/Desktop/test.pdf but interestingly enough the rectangle in pdf is empty with a black border...

But this test was on 10.13.5 with Apache 2.2.34 via MAMP.

My native apache is 2.4.33... so I will have to setup a VM to create the exact conditions.

Any reason you aren't running 10.13.5?  Also do you have HomeBrew (https://brew.sh/) installed?

**Comment by alloylab** [comment]
Here is crash report for reference... I need to review it.

[wkhtmltopdf_Mac-Pro-crash.txt](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2107499/wkhtmltopdf_Mac-Pro-crash.txt)

**Comment by tlode** [comment]
Thank you for looking into it.

I can't imagine that it depends that much on the exact setup. 3 years ago the issue could be reproduced on different macos'es, and on a debian server, too.

> Displaying `<image src="test-pattern.svg">` works with no seg fault

but it does not render the pattern at all... at least in my case

> Also do you have HomeBrew (https://brew.sh/) installed?

yes, I do

> Any reason you aren't running 10.13.5? 

Good question :)

**Comment by alloylab** [comment]
@tlode `<image src="test-pattern.svg">` was just a note to myself :)  I agree it doesnt render the pattern correctly

**Comment by sondre-coretrek** [comment]
Hi, has this been fixed in later releases?
Believe I have a similar issue where the pattern is not rendered correctly.
The pattern becomes stretched as it tries to fill the image actual size (150px/150px) when the `<image>` has a width and height of 20

```html
<defs>
    <pattern id="pattern" x="0" y="0" width="20" height="20">
        <image width="20" height="20" xlink:href="/file.svg" preserveAspectRatio="none"></image>
    </pattern>
</defs>
<svg width="20" height="200" x="0" y="0">
  <rect width="100%" height="100%" fill="url(#pattern)"></rect>
</svg>
```
Expected (On page):
![image](https://user-images.githubusercontent.com/53086713/145398580-268fdf64-64a5-473e-addb-9a0e7436c37c.png)
Stretched (In PDF):
![image](https://user-images.githubusercontent.com/53086713/145398706-b5bbffd9-4e50-469a-93af-5c794d466fb0.png)
